### PR TITLE
Use qualified name for conflicting simple names in the same file

### PR DESCRIPTION
**Pull request by auke-**
Fix for #298

### CODE REVIEW

File: src/test/java/com/squareup/javapoet/JavaFileTest.java [hunk 1]
```diff
@@ -154,6 +152,82 @@
...
+            .build())
+        .build()
+        .toString();
+    assertThat(source).isEqualTo(""
+        + "package com.squareup.tacos;\n"
+        + "\n"
+        + "class A {\n"
+        + "  class B {\n"
+        + "    class C {\n"
+        + "      A.Twin.D d;\n"
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: Shouldn't this just be Twin?

Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: err, Twin.D ?

Comment (auke-) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: I stared for hours at this problem (and first at my generated code, why it did't compile) so i might have made a mistake here. The nested-child case seems unnecessary.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 2]
```diff
@@ -332,12 +332,36 @@ private int commonPrefixLength(List<String> classNames) {
     for (int i = 0; i < size; i++) {
       String a = classNames.get(i);
       String b = typeSpecStack.get(i).name;
-      if (!a.equals(b)) return i;
+      if (!a.equals(b)) {
+        for (; i > 0; i--) {
+          // Does the current type have a parent type with a member type with the same name
+          if (!parentTypeHasMemberType(classNames.get(i), i)) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Looks wrong. My guess is you want to invert this: if any parent has a member type return 0; if none do return i.

Comment (auke-) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I don't think this is wrong, I'm trying to find the least-qualified non-conflicting name here.

I changed the comment and commit description to make that more clear.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 3]
```diff
@@ -332,12 +332,37 @@ private int commonPrefixLength(List<String> classNames) {
     for (int i = 0; i < size; i++) {
       String a = classNames.get(i);
       String b = typeSpecStack.get(i).name;
-      if (!a.equals(b)) return i;
+      if (!a.equals(b)) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I think we need another method. This method no longer does what it's named to do, nor what it's documented to do. That new method should delegate to `commonPrefixLength()`.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 2 above)
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Ahh, got it. The import rules are hard for me to get my head around!

**Comment by auke-** [comment]
Is the latest version of the patch good enough?

**Comment by swankjesse** [comment]
Much apologies for the delay on this. The logic is subtle & I wanted to take my time reviewing it. Seems solid. Thanks so much for the fix.

**Comment by auke-** [comment]
Thanks, no problem. I think the logic can be made more straightforward by using a ClassName instance instead of a list of strings and an index and compare it to a ClassName instance constructed from the typeSpecStack.

Furthermore, the algorithm could be made less greedy. The same goes for the method conflictsWithLocalName, which returns a fully qualified name for a nested class if the outer class has a name conflict.

I should read the specs (https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html) to understand all the naming rules...

**Comment by swankjesse** [comment]
Yep. If you want to send another PR to simplify further, that's always welcome!

**Pull request merged**